### PR TITLE
Show command line in `ninja -v` for `capture: true` custom targets and generators

### PR DESCRIPTION
**Pull request by bonzini**
Using `capture: true` with `custom_target` makes "ninja -v" less than useful since the command line is completely hidden from the user. It would be nice if `meson --internal exe` knew to print the command line itself; unfortunately support for a `NINJAFLAGS` command line variable, which would pass -v down to `meson --internal exe`, was rejected.

Instead, this pull request makes `meson_exe` able to create an `ExecutableSerialisation` object from the command line, and uses this new invocation method for the common case of `capture: true` in custom targets and generators.  The code around `ExecutableSerialisation` needs some refactoring in order to support this, which in turn enables simplification of the `meson_exe` script.

### CODE REVIEW

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -69,17 +69,13 @@ def __init__(self, fname, outdir, aliases, strip, install_name_mappings, install
         self.optional = optional
 
 class ExecutableSerialisation:
-    def __init__(self, name, fname, cmd_args, env, is_cross, exe_wrapper,
-                 workdir, extra_paths, capture, needs_exe_wrapper: bool):
-        self.name = name
-        self.fname = fname
+    def __init__(self, cmd_args, env={}, exe_wrapper=None,
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__ [B006] Do not use mutable data structures for argument defaults. All calls reuse one instance of that data structure, persisting changes between them. ([link](https://sider.review/gh/repos/19784232/pulls/5644#issue-4339556))

<sub>posted by Sider</sub>

File: mesonbuild/backend/backends.py [hunk 2]
```diff
@@ -323,26 +319,59 @@ def _flatten_object_list(self, target, objects, proj_dir_to_build_root):
                 raise MesonException('Unknown data type in object list.')
         return obj_list
 
-    def serialize_executable(self, tname, exe, cmd_args, workdir, env=None,
-                             extra_paths=None, capture=None):
+    def as_meson_exe_cmdline(self, tname, exe, cmd_args, for_machine=MachineChoice.BUILD,
+                             extra_bdeps=[], capture=None):
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__ [B006] Do not use mutable data structures for argument defaults. All calls reuse one instance of that data structure, persisting changes between them. ([link](https://sider.review/gh/repos/19784232/pulls/5644#issue-4339557))

<sub>posted by Sider</sub>

File: mesonbuild/backend/backends.py [hunk 3]
```diff
@@ -323,26 +319,59 @@ def _flatten_object_list(self, target, objects, proj_dir_to_build_root):
...
+                raise MesonException(msg.format(exe_wrapper.name, tname))
+        else:
+            if exe_cmd[0].endswith('.jar'):
+                exe_cmd = ['java', '-jar'] + exe_cmd
+            elif exe_cmd[0].endswith('.exe') and not (mesonlib.is_windows() or mesonlib.is_cygwin()):
+                exe_cmd = ['mono'] + exe_cmd
+            exe_wrapper = None
+
+        if not force_serialize and not extra_paths and not exe_wrapper and
+           not any('\n' in c for c in cmd_args):
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__ [E111] indentation is not a multiple of four ([link](https://sider.review/gh/repos/19784232/pulls/5644#issue-4339602))

<sub>posted by Sider</sub>

Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__ [E113] unexpected indentation ([link](https://sider.review/gh/repos/19784232/pulls/5644#issue-4339603))

<sub>posted by Sider</sub>

File: mesonbuild/backend/backends.py [hunk 4]
```diff
@@ -323,26 +319,59 @@ def _flatten_object_list(self, target, objects, proj_dir_to_build_root):
...
+                      'check the command and/or add it to PATH.'
+                raise MesonException(msg.format(exe_wrapper.name, tname))
+        else:
+            if exe_cmd[0].endswith('.jar'):
+                exe_cmd = ['java', '-jar'] + exe_cmd
+            elif exe_cmd[0].endswith('.exe') and not (mesonlib.is_windows() or mesonlib.is_cygwin()):
+                exe_cmd = ['mono'] + exe_cmd
+            exe_wrapper = None
+
+        if not force_serialize and not extra_paths and not exe_wrapper and
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__ [E999] SyntaxError: invalid syntax ([link](https://sider.review/gh/repos/19784232/pulls/5644#issue-4339601))

<sub>posted by Sider</sub>

**Comment by bonzini** [comment]
No idea what is going on with CI, but the affected tests pass in the vc2015x86vs job.

**Comment by jpakkane** [comment]
Test failures seem relevant:

```
CustomBuild:
  Performing Custom Build Tools
  Traceback (most recent call last):
    File "D:\a\1\s\test cases\common\130 configure file in generator\src\gen.py", line 8, in <module>
      with open(ifile, 'r') as f:
  FileNotFoundError: [Errno 2] No such file or directory: '../inc/confdata'
  Performing Custom Build Tools
  Traceback (most recent call last):
    File "D:\a\1\s\test cases\common\130 configure file in generator\src\gen.py", line 8, in <module>
      with open(ifile, 'r') as f:
  FileNotFoundError: [Errno 2] No such file or directory: '../../test cases/common/130 configure file in generator/src/source'
```

This may be caused by the fact that the working directory is different in VS as opposed to Ninja. In the former the working directory is the subdirectory where the generated `.vcxproj` is. There is a method somewhere to calculate the path from current VS dir to build dir root, you need to call that and then chain that to the actual file to be processed.

**Comment by bonzini** [comment]
Yeah, I had just reached a similar conclusion. Thanks!

**Comment by jpakkane** [comment]
The test seems identical to `srcgen3.py` in test case `95 gen extra`.

**Comment by bonzini** [comment]
True, I removed the test now.

**Pull request merged**